flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/SlotPoolImpl.java

```diff
@@ -112,6 +114,9 @@
 	/** The requests that are waiting for the resource manager to be connected. */
 	private final LinkedHashMap<SlotRequestId, PendingRequest> waitingForResourceManager;
 
+	/** Maps a request to its allocation. */
+	private final BiMap<SlotRequestId, AllocationID> requestedAllocations;
```

Comment:
Looking into the implementation of `DualKeyLinkedMap` for `waitingForResourceManager`, it seems we can just remove the first matching `SlotRequestId` and then remap the orphaned `SlotRequestId` to its `AllocationID`. The original insertion ordering should not suffer in `DualKeyLinkedMap.aMap`. If so, we could remove  `requestedAllocations`.

Reply:
Yes we can improve `DualKeyLinkedMap` and drop the `requestedAllocations`. 
The needed improvement would be:
1. re-insert a record does not affect its order
2. get keyA from keyB, and vice versa

Reply:
done.